In [1]:
import networkx as nx
import community as community_louvain
from sklearn.metrics import normalized_mutual_info_score
import matplotlib.pyplot as plt

# On importe un graphe pour chacune des université
G_caltech = nx.read_gml('./db/Caltech36.gml')
G_mit = nx.read_gml('./db/MIT8.gml')
G_hopkins = nx.read_gml('./db/Johns Hopkins55.gml')

# Question 6
On nous demande ici de formuler une hypothèse, puis de la mettre à l'épreuve

On émet donc l'hypothèse que "les étudiants vivant dans le même dortoir sont plus susceptibles de former des communautés fortement connectées, comparativement aux étudiants partageant simplement une même majeure"

In [5]:
def evaluate_community_detection(G, attribute):
    # On détecte les communautés avec Louvain
    partition = community_louvain.best_partition(G)
    
    # On regroupe les labels réels (e.g., dortoir ou majeure)
    true_labels = []
    for n in G.nodes():
        true_labels.append(G.nodes[n].get(attribute, 'Unknown'))
    
    # On récupère les labels prédits
    pred_labels = [partition[n] for n in G.nodes()]
    
    # Comparaison via NMI
    nmi = normalized_mutual_info_score(true_labels, pred_labels)
    return nmi

# Résultats pour Caltech
nmi_dorm = evaluate_community_detection(G_caltech, 'dorm')
nmi_major = evaluate_community_detection(G_caltech, 'second_major')

print("NMI communauté vs dortoir pour Caltech :", round(nmi_dorm, 3))
print("NMI communauté vs majeure pour Caltech :", round(nmi_major, 3))

# Résultats pour MIT
nmi_dorm = evaluate_community_detection(G_mit, 'dorm')
nmi_major = evaluate_community_detection(G_mit, 'second_major')

print("NMI communauté vs dortoir pour MIT :", round(nmi_dorm, 3))
print("NMI communauté vs majeure pour MIT :", round(nmi_major, 3))


NMI communauté vs dortoir pour Caltech : 0.534
NMI communauté vs majeure pour Caltech : 0.104
NMI communauté vs dortoir pour MIT : 0.238
NMI communauté vs majeure pour MIT : 0.036


Les résultats suggèrent que les communautés détectées sont davantage alignées avec les groupes de dortoirs qu'avec les majeures des étudiants, pour Caltech (0.534 > 0.104) comme pour MIT (0.238 > 0.036).
D'où notre hypothèse selon laquelle les communautés sociales seraient mieux représentées par les dortoirs semble être vérifiée !